File: android/app/src/main/java/com/b_one/aether/security/PeerTrust.kt
package com.b_one.aether.security

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKeys
import org.json.JSONObject
import java.security.MessageDigest
import java.util.Base64

enum class PeerTrustMode {
    TOFU,
    QR_PINNED
}

data class PeerPin(
    val peerId: String,
    val publicKeyHex: String,
    val publicKeySha256: String,
    val protocolVersion: String,
    val trustMode: PeerTrustMode,
    val addedAtEpochMs: Long,
    val lastValidatedAtEpochMs: Long
)

data class PeerTrustDecision(
    val pin: PeerPin,
    val trustEstablishedNow: Boolean
)

object PeerTrust {
    const val ONBOARDING_SCHEME = "aether-peer-pin-v1"
    private const val ONBOARDING_URI_PREFIX = "aether://peer-pin?data="

    fun fingerprintHex(publicKeyBytes: ByteArray): String =
        MessageDigest.getInstance("SHA-256")
            .digest(publicKeyBytes)
            .joinToString("") { "%02x".format(it) }

    fun createOnboardingPayload(
        peerId: String,
        publicKeyHex: String,
        protocolVersion: String,
        canonicalize: (Map<String, Any?>) -> String = ::defaultCanonicalize
    ): String {
        val normalizedKeyHex = publicKeyHex.lowercase()
        val fingerprint = fingerprintHex(hexToBytes(normalizedKeyHex))
        return canonicalize(
            mapOf(
                "peer_id" to peerId,
                "protocol_version" to protocolVersion,
                "public_key_hex" to normalizedKeyHex,
                "public_key_sha256" to fingerprint,
                "scheme" to ONBOARDING_SCHEME
            )
        )
    }

    private fun defaultCanonicalize(map: Map<String, Any?>): String {
        val keys = map.keys.sorted()
        val pairs = keys.map { key ->
            val v = map.getValue(key)
            val rendered = when (v) {
                null -> "null"
                is String -> quote(v)
                is Number, is Boolean -> v.toString()
                else -> throw IllegalArgumentException("Unsupported canonical JSON value type")
            }
            "${quote(key)}:$rendered"
        }
        return "{${pairs.joinToString(",")}}"
    }

    private fun quote(s: String): String {
        val sb = StringBuilder(s.length + 2)
        sb.append('"')
        for (ch in s) {
            when (ch) {
                '\\' -> sb.append("\\\\")
                '"' -> sb.append("\\\"")
                '\b' -> sb.append("\\b")
                '\u000C' -> sb.append("\\f")
                '\n' -> sb.append("\\n")
                '\r' -> sb.append("\\r")
                '\t' -> sb.append("\\t")
                else -> if (ch.code < 0x20) sb.append("\\u%04x".format(ch.code)) else sb.append(ch)
            }
        }
        sb.append('"')
        return sb.toString()
    }

    fun createOnboardingUri(
        peerId: String,
        publicKeyHex: String,
        protocolVersion: String
    ): String {
        val payload = createOnboardingPayload(peerId, publicKeyHex, protocolVersion)
        val encoded = Base64.getUrlEncoder().withoutPadding()
            .encodeToString(payload.toByteArray(Charsets.UTF_8))
        return ONBOARDING_URI_PREFIX + encoded
    }

    fun parseOnboardingPayload(payloadOrUri: String, nowEpochMs: Long = System.currentTimeMillis()): PeerPin {
        val payload = if (payloadOrUri.startsWith(ONBOARDING_URI_PREFIX)) {
            val encoded = payloadOrUri.removePrefix(ONBOARDING_URI_PREFIX)
            String(Base64.getUrlDecoder().decode(encoded), Charsets.UTF_8)
        } else {
            payloadOrUri
        }

        val json = JSONObject(payload)
        require(json.getString("scheme") == ONBOARDING_SCHEME) { "Unsupported onboarding scheme" }

        val peerId = json.getString("peer_id")
        val publicKeyHex = json.getString("public_key_hex").lowercase()
        val expectedFingerprint = json.getString("public_key_sha256").lowercase()
        val protocolVersion = json.optString("protocol_version", "")
        val fingerprint = fingerprintHex(hexToBytes(publicKeyHex))
        require(fingerprint == expectedFingerprint) { "Onboarding payload fingerprint mismatch" }

        return PeerPin(
            peerId = peerId,
            publicKeyHex = publicKeyHex,
            publicKeySha256 = fingerprint,
            protocolVersion = protocolVersion,
            trustMode = PeerTrustMode.QR_PINNED,
            addedAtEpochMs = nowEpochMs,
            lastValidatedAtEpochMs = nowEpochMs
        )
    }

    fun evaluateHandshake(
        peerId: String,
        publicKeyHex: String,
        protocolVersion: String,
        existingPin: PeerPin?,
        trustOnFirstUse: Boolean,
        nowEpochMs: Long = System.currentTimeMillis()
    ): PeerTrustDecision {
        val normalizedKeyHex = publicKeyHex.lowercase()
        val fingerprint = fingerprintHex(hexToBytes(normalizedKeyHex))
        val currentProtocol = protocolVersion.ifBlank { existingPin?.protocolVersion.orEmpty() }

        if (existingPin != null) {
            require(existingPin.peerId == peerId) { "Pinned peer ID mismatch" }
            require(existingPin.publicKeySha256 == fingerprint) { "Pinned peer fingerprint mismatch" }
            require(existingPin.publicKeyHex == normalizedKeyHex) { "Pinned peer public key mismatch" }
            return PeerTrustDecision(
                pin = existingPin.copy(
                    protocolVersion = currentProtocol,
                    lastValidatedAtEpochMs = nowEpochMs
                ),
                trustEstablishedNow = false
            )
        }

        require(trustOnFirstUse) { "Peer is not pinned; disable TOFU only after QR onboarding" }
        return PeerTrustDecision(
            pin = PeerPin(
                peerId = peerId,
                publicKeyHex = normalizedKeyHex,
                publicKeySha256 = fingerprint,
                protocolVersion = currentProtocol,
                trustMode = PeerTrustMode.TOFU,
                addedAtEpochMs = nowEpochMs,
                lastValidatedAtEpochMs = nowEpochMs
            ),
            trustEstablishedNow = true
        )
    }

    fun hexToBytes(hex: String): ByteArray {
        require(hex.length % 2 == 0) { "Hex string must have even length" }
        return ByteArray(hex.length / 2) { i ->
            hex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }
    }
}

class PeerPinStore(context: Context) {
    // ADR-008: EncryptedSharedPreferences replaces plaintext SharedPreferences.
    // Implements the same SharedPreferences interface — read/write code unchanged.
    private val prefs: SharedPreferences = EncryptedSharedPreferences.create(
        "aether_peer_pins_encrypted",
        MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC),
        context,
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
    )

    init {
        // One-time migration: copy pins from old plaintext SharedPreferences
        // into the new encrypted store, then delete the old file.
        migrateFromPlaintext(context)
    }

    fun get(peerId: String): PeerPin? {
        val raw = prefs.getString(key(peerId), null) ?: return null
        return runCatching { decode(raw) }.getOrNull()
    }

    fun save(pin: PeerPin) {
        prefs.edit().putString(key(pin.peerId), encode(pin)).apply()
    }

    fun remove(peerId: String) {
        prefs.edit().remove(key(peerId)).apply()
    }

    private fun key(peerId: String) = "peer_pin:$peerId"

    private fun encode(pin: PeerPin): String =
        JSONObject()
            .put("peer_id", pin.peerId)
            .put("public_key_hex", pin.publicKeyHex)
            .put("public_key_sha256", pin.publicKeySha256)
            .put("protocol_version", pin.protocolVersion)
            .put("trust_mode", pin.trustMode.name)
            .put("added_at_epoch_ms", pin.addedAtEpochMs)
            .put("last_validated_at_epoch_ms", pin.lastValidatedAtEpochMs)
            .toString()

    private fun decode(raw: String): PeerPin {
        val json = JSONObject(raw)
        val peerId = json.getString("peer_id")
        val publicKeyHex = json.getString("public_key_hex")
        val publicKeySha256 = json.getString("public_key_sha256")
        val protocolVersion = json.optString("protocol_version", "")
        val trustModeStr = json.getString("trust_mode")
        val addedAtEpochMs = json.getLong("added_at_epoch_ms")
        val lastValidatedAtEpochMs = json.getLong("last_validated_at_epoch_ms")

        // Validation: bounds check on timestamps (must be reasonable epoch ms)
        val now = System.currentTimeMillis()
        val year2020 = 1577836800000L // 2020-01-01 in epoch ms
        require(addedAtEpochMs in year2020..now + 86400000L) { "Invalid added_at_epoch_ms: $addedAtEpochMs" }
        require(lastValidatedAtEpochMs in year2020..now + 86400000L) { "Invalid last_validated_at_epoch_ms: $lastValidatedAtEpochMs" }

        // Validation: public key hex must be 130 chars (65 bytes = 130 hex chars for P-256 uncompressed)
        require(publicKeyHex.length == 130 && publicKeyHex.all { it.isDigit() || it.lowercaseChar() in 'a'..'f' }) {
            "Invalid publicKeyHex format"
        }
        // Validation: SHA-256 must be 64 hex chars
        require(publicKeySha256.length == 64 && publicKeySha256.all { it.isDigit() || it.lowercaseChar() in 'a'..'f' }) {
            "Invalid publicKeySha256 format"
        }

        return PeerPin(
            peerId = peerId,
            publicKeyHex = publicKeyHex,
            publicKeySha256 = publicKeySha256,
            protocolVersion = protocolVersion,
            trustMode = PeerTrustMode.valueOf(trustModeStr),
            addedAtEpochMs = addedAtEpochMs,
            lastValidatedAtEpochMs = lastValidatedAtEpochMs
        )
    }

    private fun migrateFromPlaintext(context: Context) {
        val oldPrefs = context.getSharedPreferences("aether_peer_pins", Context.MODE_PRIVATE)
        if (oldPrefs.all.isEmpty()) return  // nothing to migrate

        val editor = prefs.edit()
        for ((k, v) in oldPrefs.all) {
            if (k.startsWith("peer_pin:") && v is String) {
                editor.putString(k, v)
            }
        }
        editor.apply()

        // Delete old plaintext data
        oldPrefs.edit().clear().apply()
    }
}

File: ios/AetherApp/Managers/PeerTrust.swift
import Foundation
import CryptoKit
import Security

enum PeerTrustMode: String, Codable {
    case tofu = "TOFU"
    case qrPinned = "QR_PINNED"
}

struct PeerPin: Codable, Equatable {
    let peerId: String
    let publicKeyHex: String
    let publicKeySha256: String
    let protocolVersion: String
    let trustMode: PeerTrustMode
    let addedAtEpochMs: UInt64
    let lastValidatedAtEpochMs: UInt64
}

struct PeerTrustDecision {
    let pin: PeerPin
    let trustEstablishedNow: Bool
}

enum PeerTrustError: LocalizedError {
    case invalidOnboardingScheme
    case fingerprintMismatch
    case pinnedPeerMismatch(String)
    case untrustedPeer

    var errorDescription: String? {
        switch self {
        case .invalidOnboardingScheme:
            return "Unsupported onboarding scheme"
        case .fingerprintMismatch:
            return "Peer fingerprint mismatch"
        case .pinnedPeerMismatch(let message):
            return message
        case .untrustedPeer:
            return "Peer is not pinned; complete QR onboarding or enable TOFU"
        }
    }
}

enum PeerTrust {
    static let onboardingScheme = "aether-peer-pin-v1"
    private static let onboardingUriPrefix = "aether://peer-pin?data="

    static func fingerprintHex(_ publicKey: Data) -> String {
        SHA256.hash(data: publicKey).map { String(format: "%02x", $0) }.joined()
    }

    static func createOnboardingPayload(
        peerId: String,
        publicKeyHex: String,
        protocolVersion: String
    ) throws -> String {
        let normalizedKeyHex = publicKeyHex.lowercased()
        guard let publicKeyData = Data(hexString: normalizedKeyHex) else {
            throw PeerTrustError.fingerprintMismatch
        }
        let payload = [
            "peer_id": peerId,
            "protocol_version": protocolVersion,
            "public_key_hex": normalizedKeyHex,
            "public_key_sha256": fingerprintHex(publicKeyData),
            "scheme": onboardingScheme
        ]
        return canonicalJSONString(payload)
    }

    static func createOnboardingURI(
        peerId: String,
        publicKeyHex: String,
        protocolVersion: String
    ) throws -> String {
        let payload = try createOnboardingPayload(
            peerId: peerId,
            publicKeyHex: publicKeyHex,
            protocolVersion: protocolVersion
        )
        let encoded = Data(payload.utf8)
            .base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
        return onboardingUriPrefix + encoded
    }

    static func parseOnboardingPayload(
        _ payloadOrURI: String,
        nowEpochMs: UInt64 = PeerPinStore.currentEpochMs()
    ) throws -> PeerPin {
        let payload: String
        if payloadOrURI.hasPrefix(onboardingUriPrefix) {
            let encoded = String(payloadOrURI.dropFirst(onboardingUriPrefix.count))
            let base64 = encoded
                .replacingOccurrences(of: "-", with: "+")
                .replacingOccurrences(of: "_", with: "/")
            let padded = base64 + String(repeating: "=", count: (4 - base64.count % 4) % 4)
            guard let data = Data(base64Encoded: padded),
                  let decoded = String(data: data, encoding: .utf8) else {
                throw PeerTrustError.invalidOnboardingScheme
            }
            payload = decoded
        } else {
            payload = payloadOrURI
        }

        guard let data = payload.data(using: .utf8),
              let raw = try JSONSerialization.jsonObject(with: data) as? [String: String],
              raw["scheme"] == onboardingScheme,
              let peerId = raw["peer_id"],
              let publicKeyHex = raw["public_key_hex"]?.lowercased(),
              let expectedFingerprint = raw["public_key_sha256"]?.lowercased(),
              let publicKeyData = Data(hexString: publicKeyHex)
        else {
            throw PeerTrustError.invalidOnboardingScheme
        }

        let fingerprint = fingerprintHex(publicKeyData)
        guard fingerprint == expectedFingerprint else {
            throw PeerTrustError.fingerprintMismatch
        }

        return PeerPin(
            peerId: peerId,
            publicKeyHex: publicKeyHex,
            publicKeySha256: fingerprint,
            protocolVersion: raw["protocol_version"] ?? "",
            trustMode: .qrPinned,
            addedAtEpochMs: nowEpochMs,
            lastValidatedAtEpochMs: nowEpochMs
        )
    }

    static func evaluateHandshake(
        peerId: String,
        publicKeyHex: String,
        protocolVersion: String,
        existingPin: PeerPin?,
        trustOnFirstUse: Bool,
        nowEpochMs: UInt64 = PeerPinStore.currentEpochMs()
    ) throws -> PeerTrustDecision {
        let normalizedKeyHex = publicKeyHex.lowercased()
        guard let publicKeyData = Data(hexString: normalizedKeyHex) else {
            throw PeerTrustError.fingerprintMismatch
        }
        let fingerprint = fingerprintHex(publicKeyData)
        let currentProtocol = protocolVersion.isEmpty ? (existingPin?.protocolVersion ?? "") : protocolVersion

        if let existingPin {
            guard existingPin.peerId == peerId else {
                throw PeerTrustError.pinnedPeerMismatch("Pinned peer ID mismatch")
            }
            guard existingPin.publicKeySha256 == fingerprint else {
                throw PeerTrustError.pinnedPeerMismatch("Pinned peer fingerprint mismatch")
            }
            guard existingPin.publicKeyHex == normalizedKeyHex else {
                throw PeerTrustError.pinnedPeerMismatch("Pinned peer public key mismatch")
            }
            return PeerTrustDecision(
                pin: PeerPin(
                    peerId: existingPin.peerId,
                    publicKeyHex: existingPin.publicKeyHex,
                    publicKeySha256: existingPin.publicKeySha256,
                    protocolVersion: currentProtocol,
                    trustMode: existingPin.trustMode,
                    addedAtEpochMs: existingPin.addedAtEpochMs,
                    lastValidatedAtEpochMs: nowEpochMs
                ),
                trustEstablishedNow: false
            )
        }

        guard trustOnFirstUse else {
            throw PeerTrustError.untrustedPeer
        }

        return PeerTrustDecision(
            pin: PeerPin(
                peerId: peerId,
                publicKeyHex: normalizedKeyHex,
                publicKeySha256: fingerprint,
                protocolVersion: currentProtocol,
                trustMode: .tofu,
                addedAtEpochMs: nowEpochMs,
                lastValidatedAtEpochMs: nowEpochMs
            ),
            trustEstablishedNow: true
        )
    }

    private static func canonicalJSONString(_ object: [String: String]) -> String {
        let keys = object.keys.sorted()
        let pairs = keys.map { key -> String in
            let value = object[key] ?? ""
            return "\(quote(key)):\(quote(value))"
        }
        return "{\(pairs.joined(separator: ","))}"
    }

    private static func quote(_ s: String) -> String {
        var out = "\""
        for scalar in s.unicodeScalars {
            switch scalar.value {
            case 0x22: out += "\\\""
            case 0x5C: out += "\\\\"
            case 0x08: out += "\\b"
            case 0x09: out += "\\t"
            case 0x0A: out += "\\n"
            case 0x0C: out += "\\f"
            case 0x0D: out += "\\r"
            case 0x00..<0x20:
                out += String(format: "\\u%04x", scalar.value)
            default:
                out += String(scalar)
            }
        }
        out += "\""
        return out
    }
}

final class PeerPinStore {
    // ADR-008: Keychain replaces UserDefaults for peer pin storage.
    // Pins are stored as kSecClassGenericPassword items with the peer ID
    // as the account name and the JSON-encoded pin as the data.
    private let service: String
    private let defaults: UserDefaults  // kept only for migration

    init(service: String = Bundle.main.bundleIdentifier ?? "com.b_one.aether",
         defaults: UserDefaults = .standard) {
        self.service = service
        self.defaults = defaults
        migrateFromUserDefaults()
    }

    func get(peerId: String) -> PeerPin? {
        loadAll()[peerId]
    }

    func save(_ pin: PeerPin) {
        var pins = loadAll()
        pins[pin.peerId] = pin
        persist(pins)
    }

    func remove(peerId: String) {
        var pins = loadAll()
        pins.removeValue(forKey: peerId)
        persist(pins)
    }

    // ── Keychain helpers ────────────────────────────────────────────────────

    private func loadAll() -> [String: PeerPin] {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitAll,
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)

        guard status == errSecSuccess, let data = result as? Data else { return [:] }
        return (try? JSONDecoder().decode([String: PeerPin].self, from: data)) ?? [:]
    }

    private func persist(_ pins: [String: PeerPin]) {
        guard let data = try? JSONEncoder().encode(pins) else { return }

        // Delete existing then add — simpler than update
        let deleteQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
        ]
        SecItemDelete(deleteQuery as CFDictionary)

        let addQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
            kSecValueData as String: data,
        ]
        SecItemAdd(addQuery as CFDictionary, nil)
    }

    // ── Migration ────────────────────────────────────────────────────────────

    private func migrateFromUserDefaults() {
        let oldKey = "aether.peerPins"
        guard let oldData = defaults.data(forKey: oldKey) else { return }

        // Read old pins
        guard let oldPins = try? JSONDecoder().decode([String: PeerPin].self, from: oldData),
              !oldPins.isEmpty else { return }

        // Merge into Keychain (existing Keychain pins take precedence)
        var existing = loadAll()
        for (id, pin) in oldPins where existing[id] == nil {
            existing[id] = pin
        }
        persist(existing)

        // Delete old UserDefaults data
        defaults.removeObject(forKey: oldKey)
    }

    static func currentEpochMs() -> UInt64 {
        UInt64(Date().timeIntervalSince1970 * 1000)
    }
}
